prov/shm/src/smr_progress.c

```diff
@@ -241,16 +249,29 @@ static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
 		FI_WARN(&smr_prov, FI_LOG_AV, "mmap error %s\n", strerror(errno));
 		return errno;
 	}
-	ret = ofi_copy_to_iov(iov, iov_count, 0, mapped_ptr, cmd->msg.hdr.size);
-	if (ret != cmd->msg.hdr.size) {
-		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-			"mmap iov copy out error\n");
-		return FI_EIO;
+	if (cmd->msg.hdr.op == ofi_op_read_req) {
+		*total_len = ofi_total_iov_len(iov, iov_count);
+		ret = ofi_copy_from_iov(mapped_ptr, *total_len, iov, iov_count, 0);
+		if (ret != *total_len) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"mmap iov copy in error\n");
+			return FI_EIO;
+		}
+	} else {
+		ret = ofi_copy_to_iov(iov, iov_count, 0, mapped_ptr,
+				      cmd->msg.hdr.size);
+		if (ret != cmd->msg.hdr.size) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"mmap iov copy out error\n");
+			return FI_EIO;
+		}
+		*total_len = ret;
 	}
+
 	*total_len = ret;
 	munmap(mapped_ptr, cmd->msg.hdr.size);
-	close(fd);
 	shm_unlink(shm_name);
+	close(fd);
```

Comment:
This should be fixed in the patch that introduces the change, or come as a separate fix

Reply:
Done.